cmd/scollector/collectors/awsBilling.go

```diff
@@ -0,0 +1,445 @@
+	}
+	awsBilling.downloader = s3manager.NewDownloader(session.New(conf)) //Gimmie a downloader
+	if awsBilling.downloader == nil {
+		return nil, fmt.Errorf("unable to create S3 downloader")
+	}
+	if purgeDays == 0 {
+		slog.Infof("S3 purging of objects is disabled")
+		awsBilling.purgeOlderThan = time.Date(2999, 12, 31, 23, 59, 59, 0, time.UTC)
+	} else {
+		purgeHours := time.Duration(-1 * 24 * purgeDays)
+		awsBilling.purgeOlderThan = time.Now().Add(purgeHours * time.Hour)
+	}
+	return awsBilling.Go()
+}
+
+type awsBillingConfig struct {
+	s3svc          *s3.S3
+	r53svc         *route53.Route53
+	downloader     *s3manager.Downloader
+	purgeOlderThan time.Time
+	bucketName     string
+	bucketPath     string
+	prodCodesReg   *regexp.Regexp
+}
+
+func (awsBilling *awsBillingConfig) Go() (opentsdb.MultiDataPoint, error) {
+	md := opentsdb.MultiDataPoint{}
+	purgeObjects := []*s3.Object{}
+	//Declare the objects we want to fetch here. For completeness sake, we're going to fetch
+	//the entire contents of the bucket, as we are going to be cleaning it out as we go.
+	getBucketObjects := &s3.ListObjectsInput{
+		Bucket: aws.String(awsBilling.bucketName),
+	}
+	//Get the objects from the bucket
+	bucketObjects, err := awsBilling.s3svc.ListObjects(getBucketObjects)
+	if err != nil {
+		return nil, err
+	}
+
+	//Sort the files found by last modified, newest first.
+	sort.Sort(sort.Reverse(ByLastModified(bucketObjects.Contents)))
+
+	//Go through the contents of the bucket and parse the different kinds of report files
+	//that that we're going to have. At the end of it we'll have a bucketContents that contains
+	//the origina S3 file, along with hopefully enough other detail.
+	var thisBucketContents []bucketContents
+	for _, bucketObject := range bucketObjects.Contents {
+		thisReport := &billingKeyStructure{}
+		thisReport.parseFromObjectKey(bucketObject.Key, awsBilling)
+		thisBucketContents = append(thisBucketContents, bucketContents{
+			origS3Item:     bucketObject,
+			awsBillingItem: thisReport,
+		})
+		//Mark files past the purge date to be purged
+		if bucketObject.LastModified.Before(awsBilling.purgeOlderThan) {
+			purgeObjects = append(purgeObjects, bucketObject)
+		}
+	}
+
+	//So, billing comes with a bunch of different manifets and whatnot. We don't really need all of that
+	//for this basic billing integration, as the first line of the CSVs we're interested in also has
+	//the breakdown of how it's all structured. We only want to process the last file that is present
+	//for each month, as each month we can get 30+ files. The most recent file for any given month should
+	//contain everything we need.
+	var monthsProcessed []string //The months that have been processed so far
+	var allBills []*billBill
+	for _, billingObject := range thisBucketContents {
+		//If the filename ends in a .gz then it's probably what we want
+		if filepath.Ext(billingObject.awsBillingItem.fileName) == ".gz" {
```

Comment:
A lof of nested Ifs, maybe better to assign some of these to bool values, and then just if a && b &&